src/referenceHandling.ts

```diff
@@ -0,0 +1,172 @@
+// (C) 2007-2019 GoodData Corporation
+import { VisualizationObject } from '@gooddata/typings';
+import { cloneDeep, get, pickBy, set } from 'lodash';
+import { v4 as uuid } from 'uuid';
+
+import {
+    IReferenceProperties,
+    IColorMappingReferenceProperty,
+    IColorMappingUriProperty,
+    IUriProperties
+} from './interfaces';
+
+type IVisualizationObject = VisualizationObject.IVisualizationObject;
+type IReferenceItems = VisualizationObject.IReferenceItems;
+
+/**
+ * Constants
+ */
+const CONTROLS_PROPERTY_PATH = ['controls'];
+const COLOR_MAPPING_PROPERTY_PATH = [...CONTROLS_PROPERTY_PATH, 'colorMapping'];
+
+/**
+ * Helpers
+ */
+const getReferenceValue = (id: string, references: IReferenceItems) => references[id];
+const getReferenceId = (value: string, references: IReferenceItems) =>
+    Object.keys(references).find(id => references[id] === value);
+
+const defaultIdGenerator = () => uuid().replace(/-/g, '');
+
+const createConverter = <TInProperties, TOutProperties>(
+    conversionBody:
+        (properties: TInProperties, references: IReferenceItems, idGenerator: () => string) => {
+            convertedProperties: TOutProperties, convertedReferences: IReferenceItems
+        }
+) => (mdObject: IVisualizationObject, idGenerator: () => string = defaultIdGenerator) => {
+    const { content } = mdObject;
+    if (!content) {
+        return mdObject;
+    }
+
+    const { properties } = content;
+    if (!properties) {
+        return mdObject;
+    }
+
+    // prepare result objects
+    const convertedMdObject: IVisualizationObject = cloneDeep(mdObject);
+    const draftProperties: TInProperties = JSON.parse(properties);
+    const draftReferences: IReferenceItems = { ...content.references } || {};
+
+    // perform the conversion
+    const { convertedProperties, convertedReferences } = conversionBody(draftProperties, draftReferences, idGenerator);
+
+    // set the new properties and references
+    convertedMdObject.content.properties = JSON.stringify(convertedProperties);
+    convertedMdObject.content.references = convertedReferences;
+
+    return convertedMdObject;
+};
+
+/**
+ * Conversion from References to URIs
+ */
+const convertColorMappingReferencesToUris = (properties: IReferenceProperties, references: IReferenceItems) => {
+    const colorMapping: IColorMappingReferenceProperty[] | undefined = get(properties, COLOR_MAPPING_PROPERTY_PATH);
+    if (!colorMapping) {
+        return;
```

Comment:
Consider robustly returning reference as uri even if reference cannot be found. Because even now the DB is contaminated by attribute value uris that are not replaced by a reference. So we will likely encounter unresolved attribute value uris that we may still use as is.

Reply:
Updated. Good catch, thanks!